@unittest.skip test error: object has no attribute '__name__'

I have a demonstration test case that produces an error on a skipped test:

    --- begin skip_bug.py ---
    import unittest
    
    class TestBug(unittest.TestCase):
        @unittest.skip
        def test_case(self):
            pass
    --- end skip_bug.py

Running pytest with --full-trace produces the following output

    ============================ test session starts =============================
    platform win32 -- Python 2.7.9, pytest-2.8.2, py-1.4.30, pluggy-0.3.1
    rootdir: c:\temp\pytest, inifile:
    collected 1 items

    skip_bug.py F
    
    ================================== FAILURES ==================================
    _____________________________ TestBug.test_case ______________________________
    
    self = <skip_bug.TestBug testMethod=test_case>
    result = <TestCaseFunction 'test_case'>
    
        def run(self, result=None):
            orig_result = result
            if result is None:
                result = self.defaultTestResult()
                startTestRun = getattr(result, 'startTestRun', None)
                if startTestRun is not None:
                    startTestRun()
    
            self._resultForDoCleanups = result
            result.startTest(self)
    
            testMethod = getattr(self, self._testMethodName)
            if (getattr(self.__class__, "__unittest_skip__", False) or
                getattr(testMethod, "__unittest_skip__", False)):
                # If the class or method was skipped.
                try:
                    skip_why = (getattr(self.__class__, '__unittest_skip_why__', '')
    
                                or getattr(testMethod, '__unittest_skip_why__', ''))
    
                    self._addSkip(result, skip_why)
                finally:
                    result.stopTest(self)
                return
            try:
                success = False
                try:
                    self.setUp()
                except SkipTest as e:
                    self._addSkip(result, str(e))
                except KeyboardInterrupt:
                    raise
                except:
                    result.addError(self, sys.exc_info())
                else:
                        try:
    >                   testMethod()
   
    C:\Python27\lib\unittest\case.py:329:
    _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
    
    test_item = <skip_bug.TestBug testMethod=test_case>
    
        def decorator(test_item):
            if not isinstance(test_item, (type, types.ClassType)):
    >           @functools.wraps(test_item)
                def skip_wrapper(*args, **kwargs):
    
    C:\Python27\lib\unittest\case.py:60:
    _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
    
    wrapper = <function skip_wrapper at 0x00000000036FCF98>
    wrapped = <skip_bug.TestBug testMethod=test_case>
    assigned = ('__module__', '__name__', '__doc__'), updated = ('__dict__',)
    
        def update_wrapper(wrapper,
                           wrapped,
                           assigned = WRAPPER_ASSIGNMENTS,
                           updated = WRAPPER_UPDATES):
            """Update a wrapper function to look like the wrapped function
        
                   wrapper is the function to be updated
                   wrapped is the original function
               assigned is a tuple naming the attributes assigned directly
               from the wrapped function to the wrapper function (defaults to
               functools.WRAPPER_ASSIGNMENTS)
               updated is a tuple naming the attributes of the wrapper that
               are updated with the corresponding attribute from the wrapped
               function (defaults to functools.WRAPPER_UPDATES)
            """
            for attr in assigned:
    >           setattr(wrapper, attr, getattr(wrapped, attr))
    E           AttributeError: 'TestBug' object has no attribute '__name__'
    
    C:\Python27\lib\functools.py:33: AttributeError
    ========================== 1 failed in 0.31 seconds ==========================

FWIW - a very messy pip freeze output:
Bottleneck==0.8.0
Cython==0.21.1
Jinja2==2.7.3
MarkupSafe==0.23
MySQL-python==1.2.5
Pillow==2.5.3
PyOpenGL==3.1.0
PyOpenGL-accelerate==3.1.0
PySide==1.2.2
PyYAML==3.11
Pygments==1.6
SQLAlchemy==0.9.7
Scipy-stack==14.8.27
Sphinx==1.2.2
Sphinx-PyPI-upload==0.2.1
VTK==5.10.1
XlsxWriter==0.5.7
Yapsy==1.10.423
backports.ssl-match-hostname==3.4.0.2
beautifulsoup4==4.3.2
blist==1.3.6
boto==2.38.0
cellh5==1.2.0
-e git+git@github.com:CellProfiler/centrosome.git@cd6b214e4803a66f863b71762eee70
5481f043e1#egg=centrosome-origin/HEAD
certifi==14.05.14
cffi==0.8.6
decorator==3.4.0
docutils==0.12
faulthandler==2.3
futures==2.1.6
gmpy2==2.0.3
greenlet==0.4.5
h5py==2.3.1
html5lib==0.999
ilastik==0.5
ipython==2.2.0
-e git+git@github.com:CellProfiler/python-javabridge.git@7c756018934f2c2c6d28d2b
12b5c9cde6a7ae55a#egg=javabridge-master
lazyflow==0.0.0
libtiff==0.4.0
lxml==3.3.5
matplotlib==1.4.0
networkx==1.9
nose==1.3.4
numexpr==2.4
numpy==1.8.2
objgraph==2.0.1
openpyxl==1.8.6
pandas==0.14.1
patsy==0.3.0
pkginfo==1.2.1
psutil==2.1.1
psycopg2==2.5.3
py==1.4.30
py2exe==0.6.9
pybrain==0.3.3
pycairo==1.10.0
pycares==0.6.1
pycosat==0.6.0
pycurl==7.19.5
pyparsing==2.0.2
pyreadline==2.0
pytest==2.8.2
-e git+git@github.com:CellProfiler/python-bioformats@ea64bbc00112e558556f7aee575
eea0780342d3a#egg=python_bioformats-origin/issues/42
python-dateutil==2.2
pytools==2014.3
pytz==2014.4
pywin32==219
pyzmq==14.3.1
qimage2ndarray==1.4
requests==2.3.0
scikit-image==0.11.3
scikit-learn==0.15.2
scipy==0.14.0
simplejson==3.6.3
six==1.7.3
statsmodels==0.5.0
sympy==0.7.5
tables==3.1.1
tornado==4.0.1
twine==1.5.0
verlib==0.1
vigranumpy==1.10.0
virtualenv==1.11.6
volumina==0.6a
wsgiref==0.1.2
wxPython==3.0.2.0
wxPython-common==3.0.2.0
xlrd==0.9.3
xlwt==0.7.5
yattag==1.5.0

--- a/AUTHORS
+++ b/AUTHORS
@@ -46,6 +46,7 @@ Jason R. Coombs
 Jurko Gospodnetić
 Katarzyna Jachim
 Kevin Cox
+Lee Kamentsky
 Maciek Fijalkowski
 Maho
 Marc Schlaich

--- a/CHANGELOG
+++ b/CHANGELOG
@@ -1,6 +1,9 @@
 2.8.3.dev
 ---------
 
+- fix #1169: add __name__ attribute to testcases in TestCaseFunction to
+  support the @unittest.skip decorator on functions and methods.
+
 - fix #1035: collecting tests if test module level obj has __getattr__().
   Thanks Suor for the report and Bruno Oliveira / Tom Viner for the PR.
 

--- a/_pytest/unittest.py
+++ b/_pytest/unittest.py
@@ -69,6 +69,16 @@ class TestCaseFunction(pytest.Function):
 
     def setup(self):
         self._testcase = self.parent.obj(self.name)
+        #
+        # See issue #1169
+        #
+        # The @unittest.skip decorator calls functools.wraps(self._testcase)
+        # The call to functools.wraps() fails unless self._testcase
+        # has a __name__ attribute. This is usually automatically supplied
+        # if the test is a function or method, but we need to add manually
+        # here.
+        #
+        setattr(self._testcase, "__name__", self.name)
         self._obj = getattr(self._testcase, self.name)
         if hasattr(self._testcase, 'setup_method'):
             self._testcase.setup_method(self._obj)
@@ -134,7 +144,6 @@ class TestCaseFunction(pytest.Function):
         pass
 
     def runtest(self):
-        setattr(self._testcase, "__name__", self.name)
         self._testcase(result=self)
 
     def _prunetraceback(self, excinfo):

--- a/_pytest/unittest.py
+++ b/_pytest/unittest.py
@@ -134,6 +134,7 @@ class TestCaseFunction(pytest.Function):
         pass
 
     def runtest(self):
+        setattr(self._testcase, "__name__", self.name)
         self._testcase(result=self)
 
     def _prunetraceback(self, excinfo):

--- a/testing/test_unittest.py
+++ b/testing/test_unittest.py
@@ -719,6 +719,7 @@ def test_unittest_raise_skip_issue748(testdir):
         *1 skipped*
     """)
 
+@pytest.mark.skipif("sys.version_info < (2,7)")
 def test_unittest_skip_issue1169(testdir):
     testpath = testdir.makepyfile(test_foo="""
         import unittest

--- a/testing/test_unittest.py
+++ b/testing/test_unittest.py
@@ -718,3 +718,16 @@ def test_unittest_raise_skip_issue748(testdir):
         *SKIP*[1]*test_foo.py*skipping due to reasons*
         *1 skipped*
     """)
+
+def test_unittest_skip_issue1169(testdir):
+    testpath = testdir.makepyfile(test_foo="""
+        import unittest
+        
+        class MyTestCase(unittest.TestCase):
+            @unittest.skip
+            def test_skip(self):
+                 self.fail()
+        """)
+    reprec = testdir.inline_run(testpath)
+    reprec.assertoutcome(passed=1)
+    

